File: hyperkit-agent/scripts/release/update-readme-links.js
#!/usr/bin/env node
/**
 * Update README.md Links for Devlog Branch Strategy
 * 
 * Converts relative links pointing to documentation in devlog branch
 * to GitHub URLs, ensuring links work from main branch (code-only).
 * 
 * This script is automatically run during version bumps to maintain
 * link integrity across branches.
 */

const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.resolve(__dirname, '../..');
const REPO_URL = 'https://github.com/JustineDevs/Hyperkit-Agent';

// Links that should point to devlog branch (convert to GitHub URLs)
const DEVLOG_LINK_PATTERNS = [
  /hyperkit-agent\/REPORTS\//,
  /hyperkit-agent\/docs\/TEAM\//,
  /hyperkit-agent\/docs\/EXECUTION\//,
  /hyperkit-agent\/docs\/INTEGRATION\//,
  /hyperkit-agent\/docs\/ROADMAP\.md/,
  /hyperkit-agent\/docs\/API_REFERENCE\.md/,
  /hyperkit-agent\/docs\/SECURITY_SETUP\.md/,
  /^docs\//,
  /^REPORTS\//,
];

// Links to keep relative (these files exist in main branch)
const KEEP_RELATIVE_PATTERNS = [
  /README\.md$/,
  /CHANGELOG\.md$/,
  /LICENSE\.md$/,
  /SECURITY\.md$/,
  /CONTRIBUTING\.md$/,
  /CODE_OF_CONDUCT\.md$/,
  /hyperkit-agent\/docs\/GUIDE\/QUICK_START\.md/,
  /hyperkit-agent\/docs\/GUIDE\/ENVIRONMENT_SETUP\.md/,
  /hyperkit-agent\/config\.yaml$/,
  /hyperkit-agent\/pyproject\.toml$/,
  /package\.json$/,
  /^VERSION$/,
];

// External URLs, anchors, and special patterns to skip
const SKIP_PATTERNS = [
  /^https?:\/\//,  // Already external URLs
  /^#/,            // Anchor links
  /^mailto:/,      // Email links
  /^ftp:\/\//,     // FTP links
];

function shouldConvertToGithubUrl(linkPath) {
  // Skip if already external URL or anchor
  for (const pattern of SKIP_PATTERNS) {
    if (pattern.test(linkPath)) {
      return false;
    }
  }
  
  // Check if it's in KEEP_RELATIVE list (exists in main)
  for (const pattern of KEEP_RELATIVE_PATTERNS) {
    if (pattern.test(linkPath)) {
      return false;
    }
  }
  
  // Check if it matches DEVLOG_LINKS patterns (should be in devlog)
  for (const pattern of DEVLOG_LINK_PATTERNS) {
    if (pattern.test(linkPath)) {
      return true;
    }
  }
  
  // Default: keep relative (might be code files or other assets)
  return false;
}

function convertToGithubUrl(linkPath) {
  // Remove leading ./ if present
  linkPath = linkPath.replace(/^\.\//, '');
  
  // Remove leading / if present
  linkPath = linkPath.replace(/^\//, '');
  
  // Convert to GitHub blob URL
  return `${REPO_URL}/blob/devlog/${linkPath}`;
}

function updateReadmeLinks(readmePath, dryRun = false) {
  if (!fs.existsSync(readmePath)) {
    console.error(`❌ README.md not found: ${readmePath}`);
    process.exit(1);
  }
  
  let content = fs.readFileSync(readmePath, 'utf8');
  const changes = [];
  
  // Pattern to match markdown links: [text](path) or [text](path "title")
  const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
  
  content = content.replace(linkPattern, (match, linkText, linkPath) => {
    // Remove title if present (e.g., "title" in [text](path "title"))
    if (linkPath.includes(' "')) {
      linkPath = linkPath.split(' "')[0];
    }
    
    // Skip if already a GitHub URL or external URL
    if (linkPath.startsWith('http')) {
      return match;
    }
    
    // Skip if it's an anchor link
    if (linkPath.startsWith('#')) {
      return match;
    }
    
    // Check if should convert
    if (shouldConvertToGithubUrl(linkPath)) {
      const newUrl = convertToGithubUrl(linkPath);
      changes.push([linkPath, newUrl]);
      return `[${linkText}](${newUrl})`;
    }
    
    return match;
  });
  
  return { content, changes };
}

function main() {
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run');
  const fileArg = args.find(arg => arg.startsWith('--file='));
  const readmePath = fileArg 
    ? path.resolve(ROOT_DIR, fileArg.split('=')[1])
    : path.join(ROOT_DIR, 'README.md');
  
  if (!fs.existsSync(readmePath)) {
    console.error(`❌ README.md not found: ${readmePath}`);
    process.exit(1);
  }
  
  const { content, changes } = updateReadmeLinks(readmePath, dryRun);
  
  if (changes.length > 0) {
    if (dryRun) {
      console.log(`[DRY RUN] Would update ${changes.length} links in ${path.relative(ROOT_DIR, readmePath)}`);
      console.log('\nChanges that would be made:');
      for (let i = 0; i < Math.min(20, changes.length); i++) {
        const [old, newUrl] = changes[i];
        console.log(`  ${old}`);
        console.log(`    -> ${newUrl}`);
      }
      if (changes.length > 20) {
        console.log(`\n  ... and ${changes.length - 20} more changes`);
      }
      console.log(`\n[INFO] Run without --dry-run to apply changes`);
    } else {
      fs.writeFileSync(readmePath, content, 'utf8');
      console.log(`[OK] Updated ${changes.length} links in ${path.relative(ROOT_DIR, readmePath)}`);
      console.log('\nSample changes:');
      for (let i = 0; i < Math.min(5, changes.length); i++) {
        const [old, newUrl] = changes[i];
        console.log(`  ${old} -> ${newUrl}`);
      }
      if (changes.length > 5) {
        console.log(`  ... and ${changes.length - 5} more`);
      }
    }
  } else {
    console.log(`[INFO] No links needed updating in ${path.relative(ROOT_DIR, readmePath)}`);
  }
}

if (require.main === module) {
  main();
}

module.exports = { updateReadmeLinks, shouldConvertToGithubUrl, convertToGithubUrl };

File: hyperkit-agent/scripts/ci/update_readme_links.py
#!/usr/bin/env python3
"""
Update README.md Links for Devlog Branch Strategy

Converts relative links pointing to documentation in devlog branch
to GitHub URLs, ensuring links work from main branch (code-only).

This script is automatically run during version bumps to maintain
link integrity across branches.
"""

import re
import sys
from pathlib import Path
from typing import List, Tuple, Optional

# Repository configuration
REPO_URL = "https://github.com/JustineDevs/Hyperkit-Agent"
REPO_OWNER = "JustineDevs"
REPO_NAME = "Hyperkit-Agent"

# Links that should point to devlog branch (convert to GitHub URLs)
DEVLOG_LINK_PATTERNS = [
    r'hyperkit-agent/REPORTS/',
    r'hyperkit-agent/docs/TEAM/',
    r'hyperkit-agent/docs/EXECUTION/',
    r'hyperkit-agent/docs/INTEGRATION/',
    r'hyperkit-agent/docs/ROADMAP\.md',
    r'hyperkit-agent/docs/API_REFERENCE\.md',
    r'hyperkit-agent/docs/SECURITY_SETUP\.md',
    r'^docs/',
    r'^REPORTS/',
]

# Links to keep relative (these files exist in main branch)
KEEP_RELATIVE_PATTERNS = [
    r'README\.md',
    r'CHANGELOG\.md',
    r'LICENSE\.md',
    r'SECURITY\.md',
    r'CONTRIBUTING\.md',
    r'CODE_OF_CONDUCT\.md',
    r'hyperkit-agent/docs/GUIDE/QUICK_START\.md',
    r'hyperkit-agent/docs/GUIDE/ENVIRONMENT_SETUP\.md',
    r'hyperkit-agent/config\.yaml',
    r'hyperkit-agent/pyproject\.toml',
    r'package\.json',
    r'VERSION',
]

# External URLs, anchors, and special patterns to skip
SKIP_PATTERNS = [
    r'^https?://',  # Already external URLs
    r'^#',          # Anchor links
    r'^mailto:',    # Email links
    r'^ftp://',     # FTP links
]


def should_convert_to_github_url(link_path: str) -> bool:
    """
    Determine if a link should be converted to GitHub URL pointing to devlog.
    
    Args:
        link_path: Relative path from the link
        
    Returns:
        True if link should be converted to GitHub URL
    """
    # Skip if already external URL or anchor
    for pattern in SKIP_PATTERNS:
        if re.match(pattern, link_path):
            return False
    
    # Check if it's in KEEP_RELATIVE list (exists in main)
    for pattern in KEEP_RELATIVE_PATTERNS:
        if re.search(pattern, link_path):
            return False
    
    # Check if it matches DEVLOG_LINKS patterns (should be in devlog)
    for pattern in DEVLOG_LINK_PATTERNS:
        if re.search(pattern, link_path):
            return True
    
    # Default: keep relative (might be code files or other assets)
    return False


def convert_to_github_url(link_path: str) -> str:
    """
    Convert relative link to GitHub blob URL pointing to devlog branch.
    
    Args:
        link_path: Relative path from the link
        
    Returns:
        GitHub blob URL pointing to devlog branch
    """
    # Remove leading ./ if present
    link_path = link_path.lstrip('./')
    
    # Remove leading / if present
    link_path = link_path.lstrip('/')
    
    # Convert to GitHub blob URL
    github_url = f"{REPO_URL}/blob/devlog/{link_path}"
    return github_url


def update_readme_links(readme_path: Path, dry_run: bool = False) -> Tuple[str, List[Tuple[str, str]]]:
    """
    Update links in README.md and return updated content and list of changes.
    
    Args:
        readme_path: Path to README.md file
        dry_run: If True, don't write changes, just return them
        
    Returns:
        Tuple of (updated_content, list_of_changes)
    """
    if not readme_path.exists():
        print(f"❌ README.md not found: {readme_path}")
        sys.exit(1)
    
    content = readme_path.read_text(encoding='utf-8')
    changes = []
    
    # Pattern to match markdown links: [text](path) or [text](path "title")
    link_pattern = r'\[([^\]]+)\]\(([^)]+)\)'
    
    def replace_link(match):
        link_text = match.group(1)
        link_path = match.group(2)
        
        # Remove title if present (e.g., "title" in [text](path "title"))
        if ' "' in link_path:
            link_path = link_path.split(' "')[0]
        
        # Skip if already a GitHub URL or external URL
        if link_path.startswith('http'):
            return match.group(0)
        
        # Skip if it's an anchor link
        if link_path.startswith('#'):
            return match.group(0)
        
        # Check if should convert
        if should_convert_to_github_url(link_path):
            new_url = convert_to_github_url(link_path)
            changes.append((link_path, new_url))
            return f'[{link_text}]({new_url})'
        
        return match.group(0)
    
    updated_content = re.sub(link_pattern, replace_link, content)
    
    return updated_content, changes


def main():
    """Main function to update README.md links."""
    import argparse
    
    parser = argparse.ArgumentParser(
        description='Update README.md links for devlog branch strategy'
    )
    parser.add_argument(
        '--dry-run',
        action='store_true',
        help='Show what would be changed without modifying files'
    )
    parser.add_argument(
        '--file',
        type=str,
        default='README.md',
        help='Path to README.md file (default: README.md)'
    )
    
    args = parser.parse_args()
    
    readme_path = Path(args.file)
    
    if not readme_path.exists():
        print(f"❌ README.md not found: {readme_path}")
        sys.exit(1)
    
    updated_content, changes = update_readme_links(readme_path, dry_run=args.dry_run)
    
    if changes:
        if args.dry_run:
            print(f"[DRY RUN] Would update {len(changes)} links in {readme_path}")
            print("\nChanges that would be made:")
            for old, new in changes[:20]:  # Show first 20
                print(f"  {old}")
                print(f"    -> {new}")
            if len(changes) > 20:
                print(f"\n  ... and {len(changes) - 20} more changes")
            print(f"\n[INFO] Run without --dry-run to apply changes")
        else:
            readme_path.write_text(updated_content, encoding='utf-8')
            print(f"[OK] Updated {len(changes)} links in {readme_path}")
            print("\nSample changes:")
            for old, new in changes[:5]:  # Show first 5
                print(f"  {old} -> {new}")
            if len(changes) > 5:
                print(f"  ... and {len(changes) - 5} more")
    else:
        print(f"[INFO] No links needed updating in {readme_path}")


if __name__ == "__main__":
    main()

